proto/scripts/protoc_plugin.go

```diff
@@ -0,0 +1,291 @@
+	if err != nil {
+		return false, false
+	}
+	if value == nil {
+		return false, false
+	}
+	if v, ok := value.(*bool); !ok || v == nil {
+		return false, false
+	} else {
+		return *v, true
+	}
+}
+
+func GetMessageBoolExtension(msg *descriptor.DescriptorProto, extension *proto.ExtensionDesc) (bool, bool) {
+	if msg.Options == nil {
+		return false, false
+	}
+	value, err := proto.GetExtension(msg.Options, extension)
+	if err != nil {
+		return false, false
+	}
+	if value == nil {
+		return false, false
+	}
+	val, ok := value.(*bool)
+	if !ok || val == nil {
+		return false, false
+	}
+	return *val, true
+}
+
+func HandleCustomExtensions(msg *descriptor.DescriptorProto) {
+	generateXmlTags := true
+	if generate, ok := GetMessageBoolExtension(msg, ext.E_XmlTags); ok {
+		generateXmlTags = generate
+	}
+
+	vanity.ForEachField([]*descriptor.DescriptorProto{msg}, func(field *descriptor.FieldDescriptorProto) {
+		if field.Options == nil {
+			field.Options = &descriptor.FieldOptions{}
+		}
+		deprecated := field.Options.Deprecated != nil && *field.Options.Deprecated == true
+
+		if field.Type != nil && *field.Type == descriptor.FieldDescriptorProto_TYPE_INT32 {
+			SetStringFieldOption(field, gogoproto.E_Casttype, "int")
+		}
+
+		if field.TypeName != nil && *field.TypeName == ".google.protobuf.Timestamp" {
+			vanity.SetBoolFieldOption(gogoproto.E_Stdtime, true)(field)
+		}
+
+		if goName, ok := GetFieldStringExtension(field, ext.E_Goname); ok {
+			SetStringFieldOption(field, gogoproto.E_Customname, goName)
+		} else if deprecated {
+			SetStringFieldOption(field, gogoproto.E_Customname, "Deprecated"+toCamelCase(*field.Name, true))
+		}
+
+		if val, ok := GetFieldBooleanExtension(field, ext.E_DeviceId); ok && val {
+			SetStringFieldOption(field, gogoproto.E_Customtype, "github.com/syncthing/syncthing/lib/protocol.DeviceID")
+		}
+
+		if jsonValue, ok := GetFieldStringExtension(field, ext.E_Json); ok {
+			SetStringFieldOption(field, gogoproto.E_Jsontag, jsonValue)
+		} else if deprecated {
+			SetStringFieldOption(field, gogoproto.E_Jsontag, "-")
+		} else {
+			SetStringFieldOption(field, gogoproto.E_Jsontag, toCamelCase(*field.Name, false))
+		}
+
+		current := ""
+		if v, ok := GetFieldStringExtension(field, gogoproto.E_Moretags); ok {
+			current = v
+		}
+
+		if generateXmlTags {
+			if len(current) > 0 {
+				current += " "
+			}
+			if xmlValue, ok := GetFieldStringExtension(field, ext.E_Xml); ok {
+				current += fmt.Sprintf(`xml:"%s"`, xmlValue)
+			} else {
+				xmlValue = toCamelCase(*field.Name, false)
+				// XML dictates element name within the collection, not collection name, so trim plural suffix.
+				if field.IsRepeated() {
+					if strings.HasSuffix(xmlValue, "ses") {
+						// addresses -> address
+						xmlValue = strings.TrimSuffix(xmlValue, "es")
+					} else {
+						// devices -> device
+						xmlValue = strings.TrimSuffix(xmlValue, "s")
+					}
+				}
+				if deprecated {
+					xmlValue += ",omitempty"
+				}
```

Comment:
holy shit the magic ✨  I didn't notice this stuff on first pass

Reply:
Magic is magical